File: examples/csharp/example_consumer/example_consumer.cs
class Program
{
    static void PrintVec(string label, Example.Const_Vector3 v)
    {
        using var s = v.ToString();
        Console.WriteLine($"{label} = {s}");
    }

    static void Main()
    {
        Example.SayHello();
        Console.WriteLine($"Square(7) = {Example.Square(7)}");

        // Dog
        using var dog = new Example.Dog("Rex", 3, "Labrador");
        using var dogSound = dog.Speak();
        Console.WriteLine((string)dogSound);
        dog.Fetch("ball");

        // Cat (indoor defaults to true when null)
        using var cat = new Example.Cat("Whiskers", 5);
        using var catSound = cat.Speak();
        Console.WriteLine((string)catSound);
        cat.Purr();

        // Polymorphism via upcast to Animal (cast to Const_Dog first to resolve ambiguity)
        Example.Const_Animal animal = (Example.Const_Dog)dog;
        using var polySound = Example.MakeAnimalSpeak(animal);
        Console.WriteLine($"Via base: {(string)polySound}");

        // Vector3 operators
        using var a = new Example.Vector3(1f, 2f, 3f);
        using var b = new Example.Vector3(4f, 5f, 6f);

        PrintVec("a", a);
        PrintVec("b", b);

        using var sum    = a + b;
        using var diff   = a - b;
        using var scaled = a * 3f;
        using var halved = a / 2f;

        PrintVec("a + b", sum);
        PrintVec("a - b", diff);
        PrintVec("a * 3", scaled);
        PrintVec("a / 2", halved);

        Console.WriteLine($"a . b = {a.Dot(b)}");
        Console.WriteLine($"|a|^2 = {a.LengthSquared()}");

        using var cross = a.Cross(b);
        PrintVec("a x b", cross);

        Console.WriteLine($"a == b: {a == b}");
        Console.WriteLine($"a != b: {a != b}");

        using var a2 = new Example.Vector3(1f, 2f, 3f);
        Console.WriteLine($"a == a2: {a == a2}");

        // Compound assignment
        a2.AddAssign(b);
        PrintVec("a2 += b", a2);
    }
}

File: examples/c/example_consumer.c
#include <example.h>
#include <std_string.h>
#include <stdio.h>

static void print_vec(const char *label, const Example_Vector3 *v)
{
    Example_std_string *s = Example_Vector3_ToString(v);
    printf("%s = %s\n", label, Example_std_string_data(s));
    Example_std_string_Destroy(s);
}

int main(void)
{
    Example_SayHello();
    printf("Square(7) = %d\n", Example_Square(7));

    /* Dog */
    Example_Dog *dog = Example_Dog_Construct("Rex", NULL, 3, "Labrador", NULL);

    Example_std_string *dog_sound = Example_Dog_Speak(dog);
    printf("%s\n", Example_std_string_data(dog_sound));
    Example_std_string_Destroy(dog_sound);

    Example_Dog_Fetch(dog, "ball", NULL);

    /* Cat (pass NULL for indoor to use the default argument = true) */
    Example_Cat *cat = Example_Cat_Construct("Whiskers", NULL, 5, NULL);

    Example_std_string *cat_sound = Example_Cat_Speak(cat);
    printf("%s\n", Example_std_string_data(cat_sound));
    Example_std_string_Destroy(cat_sound);

    Example_Cat_Purr(cat);

    /* Polymorphism via base-class pointer */
    const Example_Animal *animal = Example_Dog_UpcastTo_Example_Animal(dog);
    Example_std_string *poly_sound = Example_MakeAnimalSpeak(animal);
    printf("Via base: %s\n", Example_std_string_data(poly_sound));
    Example_std_string_Destroy(poly_sound);

    Example_Dog_Destroy(dog);
    Example_Cat_Destroy(cat);

    /* Vector3 operators */
    Example_Vector3 *a = Example_Vector3_Construct(1.0f, 2.0f, 3.0f);
    Example_Vector3 *b = Example_Vector3_Construct(4.0f, 5.0f, 6.0f);

    print_vec("a", a);
    print_vec("b", b);

    Example_Vector3 *sum    = Example_add_Example_Vector3(a, b);
    Example_Vector3 *diff   = Example_sub_Example_Vector3(a, b);
    Example_Vector3 *scaled = Example_mul_Example_Vector3_float(a, 3.0f);
    Example_Vector3 *halved = Example_div_Example_Vector3_float(a, 2.0f);

    print_vec("a + b", sum);
    print_vec("a - b", diff);
    print_vec("a * 3", scaled);
    print_vec("a / 2", halved);

    printf("a . b = %f\n", Example_Vector3_Dot(a, b));
    printf("|a|^2 = %f\n", Example_Vector3_LengthSquared(a));

    Example_Vector3 *cross = Example_Vector3_Cross(a, b);
    print_vec("a x b", cross);

    printf("a == b: %d\n", Example_equal_Example_Vector3(a, b));
    printf("a != b: %d\n", Example_not_equal_Example_Vector3(a, b));

    Example_Vector3 *a2 = Example_Vector3_Construct(1.0f, 2.0f, 3.0f);
    printf("a == a2: %d\n", Example_equal_Example_Vector3(a, a2));

    /* compound assignment: a2 += b */
    Example_Vector3_add_assign(a2, b);
    print_vec("a2 += b", a2);

    Example_Vector3_Destroy(a);
    Example_Vector3_Destroy(b);
    Example_Vector3_Destroy(a2);
    Example_Vector3_Destroy(sum);
    Example_Vector3_Destroy(diff);
    Example_Vector3_Destroy(scaled);
    Example_Vector3_Destroy(halved);
    Example_Vector3_Destroy(cross);

    return 0;
}
